### PR TITLE
gui/liquids: add river sources, setting/adding/removing from liquid levels, and cleaning flags

### PR DESCRIPTION
**Pull request by vallode**
Adds a few of the older features of the liquids plugin back into the GUI.

Wondering about the best way to revert setting a river source, right now the tool doesn't provide anything. Once you set it, it's there until you remove it with something like tiletypes.

About `modtools/spawn-liquids` resetting temperature... I noticed that the old liquids plugin did this everywhere but was wondering if we maybe put a boolean on that? Is there a good scenario in which someone might _not_ want to reset temperatures (they'd be updated pretty quickly regardless?)?

### CODE REVIEW

**Comment by myk002** [comment]
Unless we have a solid reason to *not* set the temp, I think we should set it. Otherwise it's just lost fps while the temp stabilizes

File: docs/gui/liquids.rst [hunk 1]
```diff
@@ -5,13 +5,14 @@ gui/liquids
     :summary: Interactively paint liquids onto the map.
     :tags: fort armok map
 
-This tool is a gui for spawning water or magma. Allowing you to click and paint
-liquid onto the map.
+This tool is a gui for spawning water, magma, and river sources on the map.
+Allowing you to click and paint liquid onto the map. It can optionally clean
```
Comment (myk002) [inline docs/gui/liquids.rst:9]: This appears to be a sentence fragment. Perhaps combine with the previous sentence?

**Comment by vallode** [comment]
> Unless we have a solid reason to _not_ set the temp, I think we should set it. Otherwise it's just lost fps while the temp stabilizes

I see that makes total sense, in that case I leave it as it is with the temperature being reset always.

File: gui/liquids.lua [hunk 2]
```diff
@@ -34,71 +43,112 @@ function SpawnLiquid:init()
...
+                { label = "Drag ", value = SpawnLiquidPaintMode.DRAG, pen = COLOR_WHITE },
+            },
+            initial_option = 1,
+            on_change = function(new, _) self.paint_mode = new end,
+        },
+        widgets.CycleHotkeyLabel{
+            frame = {l = 18, b = 2},
+            label = 'Mode:',
+            auto_width = true,
+            key = 'CUSTOM_E',
```
Comment (myk002) [inline gui/liquids.lua:94]: If you'd like to enable movement key passthrough, avoid wasdec as hotkeys

File: gui/liquids.lua [hunk 3]
```diff
@@ -178,6 +232,7 @@ end
 SpawnLiquidScreen = defclass(SpawnLiquidScreen, gui.ZScreen)
 SpawnLiquidScreen.ATTRS {
     focus_path='spawnliquid',
+    pass_movement_keys=true,
```
Comment (myk002) [inline gui/liquids.lua:235]: You probably also want to set `pass_mouse_clicks=false` and `force_pause=true` since you're affecting the map

**Comment by myk002** [comment]
> Wondering about the best way to revert setting a river source, right now the tool doesn't provide anything. Once you set it, it's there until you remove it with something like tiletypes.

funny you should bring up tiletypes. I just added a Lua API for tiletypes that you can use. Docs here: https://docs.dfhack.org/en/latest/docs/dev/Lua%20API.html#tiletypes and example usage here: https://github.com/DFHack/scripts/blob/master/build-now.lua#L476-L479

**Comment by myk002** [comment]
did you want to do the tiletypes integration in this PR, or should that be for the future?

**Comment by vallode** [comment]
> did you want to do the tiletypes integration in this PR, or should that be for the future?

Good question... I can replace the existing river source spawning mechanic with the tiletypes functions but I'm not sure about how we would even handle an undo process in this tool.

**Comment by myk002** [comment]
it's ok to leave undo out, I think. eventually we'll have `gui/tiletypes` to make that kind of cleanup easier

File: changelog.txt [hunk 4]
```diff
@@ -30,6 +30,8 @@ that repo.
 - `gui/quickfort`: don't close the window when applying a blueprint so players can apply the same blueprint multiple times more easily
 - `locate-ore`: now only searches revealed tiles by default
 - `unforbid`: does not unforbid unreachable items by default anymore.
+- `modtools/spawn-liquid`: resets temperature to "normal" levels, when spawning water or magma
```
Comment (myk002) [inline changelog.txt:33]:
```suggestion
- `modtools/spawn-liquid`: sets tile temperature to stable levels when spawning water or magma
```

File: docs/gui/liquids.rst [hunk 5]
```diff
@@ -5,13 +5,14 @@ gui/liquids
...
-liquid onto the map.
+This tool is a gui for spawning water, magma, and river sources on the map,
+allowing you to click and paint liquid onto the map. It can optionally clean
+existing liquids on the map from salt or being stagnant.
 
 .. warning::
 
     There is **no undo support**. Be sure you know what you are doing before
-    spawning any liquid!
+    spawning any liquid, especially river sources.
```
Comment (myk002) [inline docs/gui/liquids.rst:15]:
```suggestion
    spawning any liquid, especially river sources!
```

Comment (myk002) [reply docs/gui/liquids.rst:15]: : )

**Review by myk002 — APPROVED** [review]

**Pull request merged**